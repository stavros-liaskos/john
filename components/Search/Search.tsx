import { SearchProps } from './Search.types';
import { useState, SyntheticEvent } from 'react';

const Search = ({ i18n, handleSearch }: SearchProps) => {
  const [input, setInput] = useState<string>();
  if (!i18n || !i18n.button || !i18n.label) {
    return null;
  }

  const handleSubmit = (e: SyntheticEvent) => {
    e.preventDefault();
    handleSearch(input);
  };

  return (
    <div className="flex justify-center items-center h-48 bg-red-300">
      <form className="flex justify-center items-stretch h-10" noValidate onSubmit={handleSubmit}>
        <input
          className="rounded mr-4 px-2 max-w-md"
          type="text"
          name="search"
          placeholder={i18n.label}
          onChange={e => setInput(e.target.value)}
        />
        <button
          className="py-2 px-3 bg-indigo-500 text-white text-sm font-semibold rounded-md shadow focus:outline-none"
          type="submit"
        >
          {i18n.button}
        </button>
      </form>
    </div>
  );
};

export default Search;
