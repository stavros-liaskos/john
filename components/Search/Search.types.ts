export interface SearchI18n {
  button: string;
  label: string;
}

export interface SearchProps {
  i18n: SearchI18n;
  handleSearch: Function;
}
