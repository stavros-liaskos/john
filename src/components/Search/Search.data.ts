import { SearchI18n } from './Search.types';

export const searchI18n: SearchI18n = {
  label: 'Here goes the artist...',
  button: 'Search',
  searchList: {
    btnTxt: 'Follow',
    noArtists: 'No artists found',
  },
};
