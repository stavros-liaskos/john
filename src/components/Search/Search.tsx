import { SearchProps } from './Search.types';
import React, { useState } from 'react';
import Button from '../Button/Button';
import { components } from '../../types/schema';
import FormInput from '../FormInput/FormInput';
import ArtistsList from '../ArtistsList/ArtistsList';
import followArtist from '../../utils/followArtist';
import { useArtistsListContext } from '../../contexts/ArtistsList/ArtistsListContext';
import Close from '../Icons/close';
import HandGlass from '../Icons/handGlass';

const Search: React.FunctionComponent<SearchProps> = ({ i18n }) => {
  const { getFollowedArtists } = useArtistsListContext();
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const [results, setResults] = useState<components['schemas']['SearchResultArtistDto'][] | null>(null);
  const [disabled, setDisabled] = useState<boolean>(false);

  if (!i18n?.label) {
    return null;
  }
  return (
    <div className="relative flex lg:justify-center items-center flex-none h-16 md:h-20 md:border-b-2 rr-border w-full">
      <FormInput handleAction={handleSearch} i18n={i18n} actionEventTrigger={'onSubmit'}>
        {results && (
          <button onClick={() => setResults(null)}>
            <Close />
          </button>
        )}
        <Button className="btn-large md:ml-52" type="submit" disabled={disabled} loading={disabled} aria-label="search">
          <HandGlass />
        </Button>
      </FormInput>

      {results && (
        <div className="absolute px-3 bg-slate-100 dark:bg-gh-darkly border-2 rr-border top-14 md:top-16 w-full z-10">
          <ArtistsList
            i18n={i18n.searchList}
            artistsList={results}
            onButtonClick={artist => handleFollow(artist)}
            artistLoading={artistLoading}
          />
        </div>
      )}
    </div>
  );

  async function handleSearch(inputValue: string) {
    inputValue &&
      (await fetch(`/artist/search?${new URLSearchParams({ pattern: inputValue })}`, {
        method: 'GET',
        credentials: 'include',
      })
        .then(res => res.json())
        .then(result => {
          getFollowedArtists();
          return setResults(result.artists);
        })
        .catch(console.error));
  }

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);
    setDisabled(true);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
      setResults(null);
      setDisabled(false);
    };
    await followArtist(artist, finallyCb);
  }
};
Search.whyDidYouRender = true;
export default Search;
