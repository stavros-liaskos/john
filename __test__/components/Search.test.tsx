import React from 'react';
import { getByRole, getByTestId, getByText, render } from '@testing-library/react';
import Search from '../../components/Search/Search';
import { searchI18n } from '../../components/Search/Search.data';

describe('Search', () => {
  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<Search />);
  });

  it('renders with data', () => {
    const component = render(<Search i18n={searchI18n} />);
    const searchBtn = getByRole(component.container, 'button');

    expect(searchBtn).toHaveTextContent('Search');
    expect(component).toMatchSnapshot();
  });
});
