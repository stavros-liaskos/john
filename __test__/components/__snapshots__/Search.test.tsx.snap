// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders with data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex justify-center items-center h-48 bg-red-300"
      >
        <form
          class="flex justify-center items-stretch h-10"
          novalidate=""
        >
          <input
            class="rounded mr-4 px-2 max-w-md"
            name="search"
            placeholder="Here goes the artist..."
            type="text"
          />
          <button
            class="py-2 px-3 bg-indigo-500 text-white text-sm font-semibold rounded-md shadow focus:outline-none"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex justify-center items-center h-48 bg-red-300"
    >
      <form
        class="flex justify-center items-stretch h-10"
        novalidate=""
      >
        <input
          class="rounded mr-4 px-2 max-w-md"
          name="search"
          placeholder="Here goes the artist..."
          type="text"
        />
        <button
          class="py-2 px-3 bg-indigo-500 text-white text-sm font-semibold rounded-md shadow focus:outline-none"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
